scipy/stats/_hypotests.py

```diff
@@ -663,3 +666,361 @@ def somersd(x, y=None):
+        enaugh to reach good precision. More points comes at performance cost.
+
+    Returns
+    -------
+    ber : BarnardExactResult
+        The object has attributes `statistic` and `pvalue`. `statistic` is
+        the Wald's statistic with pooled or unpooled variance, depending on
+        the user choice of `pooled` param. `pvalue` is the probability of
+        obtaining a test statistic at least as extreme as the one that was
+        actually observed, assuming that the null hypothesis is true.
+
+    See Also
+    --------
+    chi2_contingency : Chi-square test of independence of variables in a
+        contingency table.
+    fisher_exact : Fisher exact test on a 2x2 contingency table.
+
+    Notes
+    -----
+    Barnard's test is an exact test used in the analysis of contingency
+    tables. It examines the association of two categorical variables and
+    is a more powerful alternative than Fisher's exact test
+    for 2x2 contingency tables.
+    When using Barnard exact test, we can assert three different null
+    hypothesis :
+
+    - :math:`H_0 : p_1 \geq p_2` versus :math:`H_1 : p_1 < p_2`,
+      with `alternative` = "less"
+
+    - :math:`H_0 : p_1 \leq p_2` versus :math:`H_1 : p_1 > p_2`,
+      with `alternative` = "greater"
+
+    - :math:`H_0 : p_1 = p_2` versus :math:`H_1 : p_1 \neq p_2`,
+      with `alternative` = "two-sided" (default one)
+
+    In order to compute Barnard's exact test, we are using the Wald
+    statistic [3]_ with pooled and unpooled variance.
+    Under the assumption that both variances are equals, use pooled variance
+    (``pooled = True``). Otherwise, use the unpooled variance (``pooled =
+    False``). With pooled variance, the statistic is computed as follow:
+
+    .. math::
+
+        T(X) = \frac{
+            \hat{p_1} - \hat{p_2}
+        }{
+            \sqrt{
+                \hat{p}(1 - \hat{p})
+                (\frac{1}{c_1} +
+                \frac{1}{c_2})
+            }
+        }
+
+    and with unpooled variance :
+
+    .. math::
+
+        T(X) = \frac{
+            \hat{p_1} - \hat{p_2}
+        }{
+            \sqrt{
+                \frac{\hat{p_1}(1 - \hat{p_1})}{c_1} +
+                \frac{\hat{p_2}(1 - \hat{p_2})}{c_2}
+            }
+        }
+
+    The p-value is then computed as follow :
+
+    .. math::
+
+        pvalue =
+        \sum
+            \binom{total\_col\_1}{x_{11}}
+            \binom{total\_col\_2}{x_{12}}
+            \pi^{x_{11} + x_{12}}
+            (1 - \pi)^{total - x_{11} - x_{12}}
+
+    Where :math:`T(X) \leq T(X_0)`, if `alternative` = "less"
+    :math:`T(X) \geq T(X_0)`, if `alternative` = "greater"
+    :math:`T(X) \geq |T(X_0)|`, if `alternative` = "two-sided".
+
+    The returned p-value is the maximum p-value taken over the nuisance
+    parameter :math:`\pi`, where :math:`0 \leq \pi \leq 1`.
+
+    This function's complexity is O(n * total_c1 * total_c2), with `n` the
+    number of initial points.
```

Comment:
Please check that symbols above are defined, e.g. \hat{p_1}, c_1, total\_col\_1, x_{11}, X, X_0, total_c1. If that is cumbersome, then consider removing definitions and instead directing the reader to reference ("calculated as in [x]_ equation ___"). 

If these stay here, I would suggest defining symbols for total\_col\_1, etc..., as they currently render as:

![image](https://user-images.githubusercontent.com/6570539/109399305-f1443880-78f6-11eb-8663-befb70a35911.png)

which could be more readable in both console and rendered docs.

I should not advise on whether to remove or add LaTeX, but I would suggest making the notation for the computational complexity consistent with that above.

Reply:
Ok, I try to change it the best as I could in 8c58fc075d7a1ae0cf7657962414ae2b772b610a. Tell me if it suits you or if you would rather have other modifications. I am not very confident on how to be at the same time concise and clear...

Reply:
Would you prefer that I explain my approach to this sort of thing, or would you prefer that I submit a PR to your branch?

Reply:
You can try explaining your approach. I am always eager to learn and there is not better way to learn than to do it ourselves ;). If ever the result doesn't suit you, you will still be able to submit a PR

Reply:
> You can try explaining your approach. I am always eager to learn and there is not better way to learn than to do it ourselves ;).

Great. I agree!

I try to proofread my own documentation according to the same standards as an academic paper. When I come across a mathematical symbol, I check to see if it has already been defined (in words) or is defined shortly thereafter. If not, I add the definition. I also try to make sure that the definition can easily be understood by most users; I avoid highly definitions that require careful study of all the rest of the documentation, the literature, or the code itself before they begin to make sense.

For instance, the first mention of p_1 and p_2 are in

![image](https://user-images.githubusercontent.com/6570539/109461755-7c1c5480-7a17-11eb-8723-333d44ea6678.png)

The next mention of them is in:

![image](https://user-images.githubusercontent.com/6570539/109461808-8c343400-7a17-11eb-8549-180cd927502c.png)

But I don't think they are ever defined in general, and only much later is there an examples:

![image](https://user-images.githubusercontent.com/6570539/109461900-ab32c600-7a17-11eb-84d9-110a893e75f6.png)

So, in this case, perhaps the reader could figure out what p_1 and p_2 mean if they read much more of the documentation and generalize from the example given, but I don't think they should have to.

One might argue that it should be obvious to users of the function what p_1 and p_2 are, as they should be familiar with the test if they are going to use it. (The definitions of p_1 and p_2 are not immediately obvious to me, but perhaps other users are more educated than I am.) In that case, I would ask why we include the definitions of the statistic at all. If "for the sake of completeness", then for the sake of completeness we should define the symbols. There might be other arguments for defining the statistic but not the notation used in its definition (e.g. "users should be familiar enough with the test to know what p_1 and p_2, but they need not be familiar enough to know how the statistic is defined"), but I find those tenuous.

Standards for documentation have varied over the years, and they still vary by reviewer. But, in short, my recommendations are to either a) define in English all symbols used in mathematical statements or b) remove those mathematical statements.

Thanks for considering it!

Again, I will not make any statements about whether symbols/mathematics here should or should not be formatted with LaTeX. (I will say that I think @jb-leger and @Carreau have written some excellent code that, if adopted more widely, would make my opinions less unpopular!)

Reply:
Thank you @mdhaber for this in depth explanation ! I think I really understood what you are waiting for and I agree with what you said. For my defense, I tried to copy the general layout `fisher_exact`'s documentation which I imagine is now quite old (I could propose a PR if you want to try to modernize it a little bit)
Anyway, since I am still quite inexperienced in writing scientific documentation / article, I am afraid that it mights not be as good as wanted. Still, the commit you want is in 44e192e.

Reply:
Thanks @V0lantis, I'll take a look this weekend. 

RE: `fisher_exact`, I think you've gone beyond what is there - especially with the careful explanation of the alternative hypotheses, the more detailed example, and the notes that define the statistic. I understand why you would compare the two, but I am not, really. I am just trying to help make the notes in `barnard_exact` self-contained and _difficult for readers to misunderstand_.

Also, standards vary between times (on average, early stats documentation is not as thorough as now) and reviewers. I just happen to be the one reviewing this now!

Reply:
I think I understand it better now! Just a few thoughts I'll indicate inline.